src/datasets/builder.py

```diff
@@ -2081,3 +2079,52 @@ def _build_pcollection(pipeline):
 
         # Add the PCollection to the pipeline
         _ = pipeline | split_name >> _build_pcollection()  # pylint: disable=no-value-for-parameter max_bytes_per_shard
+
+    def as_streaming_dataset(
+        self,
+        split: Optional[str] = None,
+    ) -> Union[Dict[str, IterableDataset], IterableDataset]:
+        self._request_info_from_hf_gcs()
+        datasets = {
+            split: IterableDataset(self._get_examples_iterable_for_split(split), info=self.info, split=split)
+            for split in self.info.splits
+        }
+        if split:
+            try:
+                datasets = datasets[split]
+            except KeyError:
+                raise ValueError(f"Bad split: {split}. Available splits: {list(datasets)}")
+        if isinstance(datasets, dict):
+            datasets = IterableDatasetDict(datasets)
+        return datasets
+
+    def _get_examples_iterable_for_split(self, split: str) -> ExamplesIterable:
+        return ExamplesIterable(self._generate_examples_from_hf_gcs, {"split": split})
+
+    def _generate_examples_from_hf_gcs(self, split):
+        remote_prepared_filename = f"{self._remote_cache_dir_from_hf_gcs}/{self.name}-{split}.arrow"
```

Comment:
(nit) Beam builders may create sharded arrow files now.

It's the case for none of the datasets on GCP but if we regenerate a new one it may en up sharded.
You can check the dataset `info.shard_lengths` to know if it's sharded or not and how many shards there are.


```python
    if self.info.shard_lengths:
        num_shards = len(self.info.shard_lengths)
        urls = [
            f"f"{self._remote_cache_dir_from_hf_gcs}/{self.name}-{split}-{shard_id:05d}-of-{num_shards:05d}.arrow"
            for shard_id in range(num_shards)
        ]
    else:
        urls = [f"{self._remote_cache_dir_from_hf_gcs}/{self.name}-{split}.arrow"]
```

Reply:
As the attribute `shard_lengths` belongs to `SplitInfo` (and not `DatasetInfo`), I have refactored the method so that split is now of type `SplitInfo` instead of `str`.